### ISSUE TITLE
Discussion: pass request to handler function instead of API endpoint

### ISSUE THREAD

**Issue by Niceadam** [OvidijusParsiunas/deep-chat#12]
I am trying to create an LLM chain with state for my application. However, I can only use the request interface to talk to stateless API endpoints on server-side. It would be much more convenient if I could pass a `handleRequest` function that does any pre-processing of the message, handles state on the client and makes API calls if I so choose. I also read about interceptor functions that can intercept and modify the request, but it still demands it eventually passes to a server API endpoint. A handler function can do the job of `requestInterceptor`, `responseInterceptor` and `request`. I don't think streaming will work though. I am using SvelteKit

**Comment by OvidijusParsiunas** [comment]
Hey @Niceadam, sorry I don't think I fully understand how the `handleRequest` function would work.

- Would you mind elaborating on how it *handles state on the client and makes API calls*.
- If you know of any libraries that already do this, they would be a good reference to see what we are aiming for here.
- Also, if you could illustrate how you would intend to use it with a code example, that would be great! Thanks!

**Comment by Niceadam** [comment]
Yes, an example would help. Currently from what I understand, deepchat uses `request` prop to connect to an API endpoint.

```svelte
<script lang="ts">
  import { onMount } from 'svelte';
  onMount(async () => {
    await import("deep-chat");
  })

  let request = JSON.stringify({
    "url": "/api/chatbot",
    "method": "POST",
  });

</script>

<main>
  <deep-chat
  request={request} />
</main>
```

Instead I was suggesting a new interface using a handler function:
```svelte
<script lang="ts">
  import { onMount } from 'svelte';
  onMount(async () => {
    await import("deep-chat");
  })

  async function handleRequest({ request }) {
    // do stuff with request message
    await fetch("/api/chatbot", {method: "post"})
    // do stuff with response
  }
</script>

<main>
  <deep-chat
  handleRequest={handleRequest} />
</main>

```

**Comment by OvidijusParsiunas** [comment]
I see what you mean, so it is more about constructing the request functionality yourself rather than leaving it to Deep Chat.
Whilst I like the idea - Deep Chat's goal is to minimize the amount of work that developers need to do and provide as much of an out of the box experience as possible, so eliminating the existing methodology would be going in the wrong direction.

Nevertheless, I do not mind bringing in *another* way to connect to APIs, something that would fit your use-case. However the immediate question I have is what unique things does `handleRequest` bring that [`requestInterceptor`](https://deepchat.dev/docs/interceptors#requestInterceptor) and [`responseInterceptor`](https://deepchat.dev/docs/interceptors#responseInterceptor) do not. You mentioned that they still demand to pass the request to an API - but doesn't `handleRequest` also demand this? And if Deep Chat would call this function (when the user attempts to send a message), wouldn't the intention be to retrieve a response?

Perphaps I am misunderstanding the use-case, maybe I could upgrade [`requestInterceptor`](https://deepchat.dev/docs/interceptors#requestInterceptor) to help out?

**Comment by OvidijusParsiunas** [comment]
Just to note, I have recently upgraded [`responseInterceptor`](https://deepchat.dev/docs/interceptors#responseInterceptor) to handle asynchronous responses as there was a dev that had to do other processing before the final result could be set in the chat. I can attempt to do that in [`requestInterceptor`](https://deepchat.dev/docs/interceptors#requestInterceptor) if it would help.
Reference [here](https://github.com/OvidijusParsiunas/deep-chat/issues/6).

**Comment by OvidijusParsiunas** [comment]
Hi, I updated the [`requestInterceptor`](https://deepchat.dev/docs/interceptors#requestInterceptor) method to be able to handle async functions which should enable you to manipulate state and do other processing before the request is sent out. This is now available in version `1.1.23`. Let me know if this helps.

**Comment by OvidijusParsiunas** [comment]
Hi, there has been no activity on this issue for the past 3 days and if you don't mind I will be closing it.
Thankyou once again for the great suggestion! The goal of this and any other library is to help developers achieve their desired use cases, but the true challenge in this process is to facilitate as much as possible with smallest amount of code as possible, otherwise the library would collapse under its own complexity, hence I have to exercise rigorous caution when adding new features/capabilities to Deep Chat.
The core use cases should currently be covered by the interceptor methods, but if I will implement a feature for calling to APIs using a method like `handleRequest` I will surely re-open this issue.
Thankyou!

**Comment by Niceadam** [comment]
I Apologize for my delayed response, I am trying to connect Langchain to the deepchat component. Langchain stores memory of previous messages for context, it stores it on the client since the server is stateless. It also handles the calls to the LLM so there are no manual api calls to be made

```svelte
<script>
	const llm = new CustomLLMApi();

	// Prompt
	const prompt = ChatPromptTemplate.fromPromptMessages([
		SystemMessagePromptTemplate.fromTemplate(`You are a chatbot talking to a human`),
		new MessagesPlaceholder('history'),
		HumanMessagePromptTemplate.fromTemplate('{input}')
	]);

	const memory = new BufferMemory({ returnMessage: true, memoryKey: 'history' });
	const chain = new ConversationChain({ llm: llm, prompt: prompt, memory: memory, verbose: true });

	async function handleRequest(request) {
		return await chain.call({ input: request.message });
	}
</script>

<DeepChat  handleRequest={handleRequest} />
```

I tried but couldn't rewrite it in terms of `requestInterceptor`, since it still requires a url, i tried setting up a dummy echo api endpoint, that just returns back the request, I do this since all my text is generated would already be generated in `requestInterceptor`  and I would already have my message from `chain.call()`

**Comment by OvidijusParsiunas** [comment]
Ok I see, so you want to defer the request logic to a different library entirely. That is an interesting use-case and probably something we ultimately need if people want to use langchain and other tools. Let me explore this for a bit and see how I can facilitate this.

**Comment by OvidijusParsiunas** [comment]
Hey, I have created a prototype version of the functionality to facilitate your requirement. It is accessible through the `deep-chat-dev` component version `9.0.0`. This component is used pretty much the same as the `deep-chat` component, except it is just using a different import name. If you don't mind, could you test it with your use-case just to make sure that it is working the way you need it before I add it to the main library.

The new functionality should be accessible via the following approach:
The `request` property accepts a new `handler` value which should be the function containing code similar to what you have described in the `handleRequest`. The difference is that the function's first argument will contain a [Request message](https://deepchat.dev/docs/connect) body and the return async object should be [`Result`](https://deepchat.dev/docs/connect#Result).

Example:
```
function async handleRequest(body) {
  const message = body.messages[0];
  let response = fetch("/api/chatbot", {
    method: "POST",
    body: {
      message: message,
      chat_history: client_store
    }
  });
  const responseText = 'response';
  return { result: {text: responseText };
}
```

One additional note, currently the component is still expecting a `url` to be defined in the request as I haven't yet configured it to not do this, so for the prototype just add a `"mock"` string to keep the component happy, e.g:

```
<deep-chat request={{url: "mock", handler: handleRequest}} />
```

Let me know if this works for and if there are any problems with this. (I will add the functionality for error handling later on, but you can assume that it will be like the one in [`responseInterceptor`](https://deepchat.dev/docs/interceptors#responseInterceptor).

**Comment by Niceadam** [comment]
Thank you for the fantastic response times. The above doesn't work for me with `url: "mock"` since it just makes a call to `http://localhost/current-url/mock` and errors out

**Comment by OvidijusParsiunas** [comment]
Hey, that's strange, please check the code in the following [live codesandbox](https://codesandbox.io/p/sandbox/deep-chat-sveltekit-forked-fhwz2q) to see if it helps.

It is very late where I live and I will be going to bed, I will be able to respond in ~10 hours. Please let me know if you can get it to work, if something doesn't seem right or opportunities for improvements. Thanks!

**Comment by Niceadam** [comment]
I am going to bed too, I will do more testing tomorrow since I still need to stringify deep-chat properties otherwise I get `Uncaught SyntaxError: JSON.parse: unexpected character `, but that is a separate issue.

**Comment by OvidijusParsiunas** [comment]
When using Sveltekit, the trick is to conditionally render the `deep-chat` when the imported module has loaded. The live code examples have been updated to reflect this as well as the example above. I go into more depth why this needs to be done in the following [issue](https://github.com/OvidijusParsiunas/deep-chat/issues/11).

**Comment by Niceadam** [comment]
I am using `deep-chat-dev: 9.0.0` and `sveltekit: 1.24.1`, I have 2 persisting problems that I tried everything but could not figure it out 1) I still need to `JSON.stringify`, else I get `Uncaught SyntaxError: JSON.parse: unexpected character` 2) `handler` gets completely ignored and makes calls to `url` anyway, I think I had the same issue with `requestInterceptor`.

```svelte
<script lang="ts">
  // Lanchain imports
	import { onMount } from 'svelte';
	import { deepChatStyles } from '$lib/static.js';

	onMount(async () => {
		await import('deep-chat-dev');
		isLoaded = true;
	});

	let isLoaded = false;

  // Langchain stuff...
	const chain = new ConversationChain({ llm: llm, prompt: prompt, memory: memory, verbose: true });

	async function handleRequest(body) {
		console.log(body);
		const response = await chain.call({ input: body.messages[0].text });
		return { result: { text: 'test' } };
	}

	const request = JSON.stringify({
		url: 'mock',
		handler: async (body) => {
			console.log('test');
			return { result: { text: 'test' } };
		}
	});

	const initialMessages = JSON.stringify([{ text: 'Hey, Welcome to test chatbot', role: 'ai' }]);
</script>

<main>
	{#if isLoaded}
		<deep-chat {...deepChatStyles} {initialMessages} {request} />
	{/if}
</main>

```

**Comment by OvidijusParsiunas** [comment]
That is strange, I have updated to `"@sveltejs/kit": "^1.24.1"` and everything appears to work fine on my end without the stringification. However I am not using the `{...deepChatStyles}` syntax which could be the problem on your end. Could you try to remove it and see if it fixes the problem?

**Comment by OvidijusParsiunas** [comment]
To anyone reading this thread and using the Deep Chat version that is `1.3.0` or higher, the result object has been replaced with [`Response`](https://deepchat.dev/docs/connect#Response).

**Comment by OvidijusParsiunas** [comment]
Based on a discussion in another [issue](https://github.com/OvidijusParsiunas/deep-chat/issues/15), I have created another handler function that can be used for streaming called `streamHandler`. This is a temporary namespace as it uses a different strategy for interacting with Deep Chat that the original `handler` does not follow and I will merge everything into the one `handler` function when I can confirm that everything works as expected and I have modified everything to use the same strategy.

What strategy am I talking about? The original `handler` function above accepts a function that would return a _Promise_ that returns a result when complete. We cannot use this approach in streaming as it expects multiple results for one request and there are other events that take place throughout its lifecycle which Deep Chat needs to be aware of. In the light of this, I have created another approach for streams:

Instead of waiting for a result, the `streamHandler` function will contain 2 arguments, the 1st will be the request message `body` (same as in the `handler`) and second will be `events` that will contain a map of functions that need to be triggered to tell the Deep Chat component what to do.

The type:

```
(
  body: any, // message body
  events: {
    onOpen: () => void; // should be called when the stream connection has been established
    onResult: (result: {text?: string; error?: string}) => void; // should be called to update streamed message or notify it about an error
    onClose: () => void; // should be called to allow the user to create a new request
    stopClicked: {listener: () => void}; // this will be triggered by the user and you can attach your own function to the listener to close the stream on your end (the component will no longer listen to the incoming messages any more, so stopping your stream when this is called is optional)
  }
) => void;
```

Example:
```
chatElement.request = {
  streamHandler: (body, events) => {
    // this is PSEUDO CODE for creating a stream using the @microsoft/fetch-event-source package and you should
    // tailor this to your streaming approach instead
    try {
      fetchEventSource('http://localhost:8080/chat-stream', {
        method: 'POST',
        body: JSON.stringify(body),
        async onopen(response) {
          if (response.ok) {
            events.onOpen; // stops the loading bubble and replaces it with a message bubble ready to receive text
          } else {
            events.onResult({error: 'Failed to connect to server'});
          }
        },
        onmessage(message) {
          events.onResult({text: message}); // adds text into the message bubble
        },
        onerror(message) {
          events.onResult({error: message}); // Deep Chat will no longer listen to this stream and display an error
        },
        onclose() {
          events.onClose(); // when stream is finished the stop button will be changed back to submit
        },
      });

      // triggered when the user clicks the stop button
      events.stopClicked.listener = () => {
        // logic to stop your stream, such as creating an abortController
      };
    } catch (e) {
      events.onResult({error: 'error'}); // Deep Chat will no longer listen to this stream and display an error
    }
  },
}
```

This is available in version `9.0.3` in `deep-chat-dev` and `deep-chat-react-dev`.

@GIOVANNI-75, let me know if you have any questions or issues when using this, thanks!

**Comment by Niceadam** [comment]
You were right, I was trying to spread props using the `{...props}` syntax but it doesn't seem to work for some reason. The `handler` function now works, Thanks!

> That is strange, I have updated to `"@sveltejs/kit": "^1.24.1"` and everything appears to work fine on my end without the stringification. However I am not using the `{...deepChatStyles}` syntax which could be the problem on your end. Could you try to remove it and see if it fixes the problem?
> 
> Also, if you copy the code in this [codesandbox](https://codesandbox.io/p/sandbox/deep-chat-sveltekit-forked-fhwz2q), does it work for you?

**Comment by GIOVANNI-75** [comment]
it doesn't work for me

**Comment by GIOVANNI-75** [comment]
It works very well, some comments: (deep-chat-dev & react-dev)
the Url argument is still necessary, requestInterceptor & onNewMessage event no longer receives data, but data recovery is still possible via the streamHandler, moreover it is simpler logically to do everything there, fetch, logic and recovery of body , it doesn't pose a problem but it was just so that you take them into account.
I can never thank you enough for your work 🙏🏻

**Comment by OvidijusParsiunas** [comment]
Happy to hear it is working for both of you!

I just looked at the code and saw that I accidentally removed the logic for the `url` to not be required (Doh!). I will release another update tomorrow where the `url` will be fixed and the `interceptors` and `onNewMessage` will also work as expected. Thankyou for letting me know!

If you experience any other issues or have any form of other feedback let me know!

**Comment by OvidijusParsiunas** [comment]
Hey, I have made a lot of changes to the `handler` API functionality which should be its final iteration that is ready to be merged into the core package.

Handler functionality is now available for basic http requests, streams and websockets using the same `handler` property. It uses a `Signals` interface which is a map of functions that will notify Deep Chat on what is happening in the `handler`.

**Type:**

```
handler?: (body: any, signals: Signals) => void;

Signals {
  onResponse: (response: Response) => void;
  onOpen?: () => void;
  onClose?: () => void;
  stopClicked?: {listener: () => void}; // exclusive for streams
  newUserMessage?: {listener: (body: any) => void}; // exclusive for websockets
}
```

The `onResponse` function uses the [`Response`](https://deepchat.dev/docs/connect#Response) type as its argument and is used for both successful and error responses.

**Basic requests**
Basic HTTP requests only use the `onResponse` signal:

```
chatElement.request = {
  handler: (body, signals) => {
    try {
      // other logic
      signals.onResponse({text: 'response text'});
    } catch (e) {
      signals.onResponse({error: 'error'});
    }
  },
}
```

**Streaming/SSE**

Make sure to have the [`stream`](https://deepchat.dev/docs/connect#stream) property set to _true_:

If you are using TypeScript, you will need to add [`optionals`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) around the signals if your linter is giving errors.

```
chatElement.request = {
  handler: (body, signals) => {
    // this is PSEUDO CODE for creating a stream using the @microsoft/fetch-event-source package and you should
    // tailor this to your streaming approach instead
    try {
      fetchEventSource('http://localhost:8080/chat-stream', {
        method: 'POST',
        body: JSON.stringify(body),
        async onopen(response) {
          if (response.ok) {
            signals.onOpen(); // stops the loading bubble and replaces it with a message bubble ready to receive text
          } else {
            signals.onResponse({error: 'Failed to connect to server'});
          }
        },
        onmessage(message) {
          signals.onResponse({text: message}); // adds text into the message bubble
        },
        onerror(message) {
          signals.onResponse({error: message}); // Deep Chat will no longer listen to this stream and display an error
        },
        onclose() {
          signals.onClose(); // when stream is finished the stop button will be changed back to submit
        },
      });

      // triggered when the user clicks the stop button
      signals.stopClicked.listener = () => {
        // logic to stop your stream, such as creating an abortController
      };
    } catch (e) {
      signals.onResponse({error: 'error'}); // Deep Chat will no longer listen to this stream and display an error
    }
  },
}
```

**Websockets**

Make sure to have the [`websocket`](https://deepchat.dev/docs/connect#Websocket) property set to _true_. This example works with the [example websocket](https://github.com/OvidijusParsiunas/deep-chat/tree/main/example-servers/node/websockets) project right out of the box. The `onOpen` property allows the user to send messages and if something happens to the connection using `onClose` prevents them from sending messages again for a better UX. Sending the `"Connection error"` error message will also display this error message inside the chat. This can be further customized using the [`errorMessages`](https://deepchat.dev/docs/messages#errorMessages) property.

If you are using TypeScript, you will need to add [`optionals`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) around the signals if your linter is giving errors.

```
chatElement.request = {
  websocket: true,
  handler: (_, signals) => {
    try {
      const websocket = new WebSocket('ws://localhost:8080');
      websocket.onopen = () => {
        signals.onOpen();
      };
      websocket.onmessage = (message) => {
        const response = JSON.parse(message.data);
        signals.onResponse(response);
      };
      websocket.onclose = () => {
        signals.onClose();
        // reconnection logic here
      };
      websocket.onerror = () => {
        signals.onResponse({error: 'error'});
        // you can also use send 'Connection error' string to display the error, also explore errorMessages property
      };
      signals.newUserMessage.listener = (body) => {
        websocket.send(JSON.stringify(body));
      };
    } catch (e) {
      signals.onResponse({error: 'error'});
      signals.onClose();
      // reconnection logic here
    }
  },
}
```

**Error Handling**

All error handling must be done inside the `handler` and respond using the `onResponse({error: message})` signal.

**Release**

This is currently available in `deep-chat-dev` and `deep-chat-react-dev` version `9.0.5`. If all goes well, I am hoping to release this in the main `deep-chat` packages tomorrow.

Please send any feedback in regards to this.

**Comment by OvidijusParsiunas** [comment]
Hey, I have made a lot of changes to the `handler` API functionality which should be its final iteration that is ready to be merged into the core package.

Handler functionality is now available for basic http requests, streams and websockets using the same `handler` property. It uses a `Signals` interface which is a map of functions that will notify Deep Chat on what is happening in the `handler`.

**Type:**

```
handler?: (body: any, signals: Signals) => void;

Signals {
  onResponse: (response: Response) => void;
  onOpen?: () => void;
  onClose?: () => void;
  stopClicked?: {listener: () => void}; // exclusive for streams
  newUserMessage?: {listener: (body: any) => void}; // exclusive for websockets
}
```

The `onResponse` function uses the [`Response`](https://deepchat.dev/docs/connect#Response) type as its argument and is used for both successful and error responses.

**Basic requests**
Basic HTTP requests only use the `onResponse` signal:

```
chatElement.request = {
  handler: (body, signals) => {
    try {
      // other logic
      signals.onResponse({text: 'response text'});
    } catch (e) {
      signals.onResponse({error: 'error'});
    }
  },
}
```

**Streaming/SSE**

Make sure to have the [`stream`](https://deepchat.dev/docs/connect#stream) property set to _true_:

If you are using TypeScript, you will need to add [`optionals`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) around the signals if your linter is giving errors.

```
chatElement.request = {
  handler: (body, signals) => {
    // this is PSEUDO CODE for creating a stream using the @microsoft/fetch-event-source package and you should
    // tailor this to your streaming approach instead
    try {
      fetchEventSource('http://localhost:8080/chat-stream', {
        method: 'POST',
        body: JSON.stringify(body),
        async onopen(response) {
          if (response.ok) {
            signals.onOpen(); // stops the loading bubble and replaces it with a message bubble ready to receive text
          } else {
            signals.onResponse({error: 'Failed to connect to server'});
          }
        },
        onmessage(message) {
          signals.onResponse({text: message}); // adds text into the message bubble
        },
        onerror(message) {
          signals.onResponse({error: message}); // Deep Chat will no longer listen to this stream and display an error
        },
        onclose() {
          signals.onClose(); // when stream is finished the stop button will be changed back to submit
        },
      });

      // triggered when the user clicks the stop button
      signals.stopClicked.listener = () => {
        // logic to stop your stream, such as creating an abortController
      };
    } catch (e) {
      signals.onResponse({error: 'error'}); // Deep Chat will no longer listen to this stream and display an error
    }
  },
}
```

**Websockets**

Make sure to have the [`websocket`](https://deepchat.dev/docs/connect#Websocket) property set to _true_. This example works with the [example websocket](https://github.com/OvidijusParsiunas/deep-chat/tree/main/example-servers/node/websockets) project right out of the box. The `onOpen` property allows the user to send messages and if something happens to the connection using `onClose` prevents them from sending messages again for a better UX. Sending the `"Connection error"` error message will also display this error message inside the chat. This can be further customized using the [`errorMessages`](https://deepchat.dev/docs/messages#errorMessages) property.

If you are using TypeScript, you will need to add [`optionals`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) around the signals if your linter is giving errors.

```
chatElement.request = {
  websocket: true,
  handler: (_, signals) => {
    try {
      const websocket = new WebSocket('ws://localhost:8080');
      websocket.onopen = () => {
        signals.onOpen();
      };
      websocket.onmessage = (message) => {
        const response = JSON.parse(message.data);
        signals.onResponse(response);
      };
      websocket.onclose = () => {
        signals.onClose();
        // reconnection logic here
      };
      websocket.onerror = () => {
        signals.onResponse({error: 'error'});
        // you can also use send 'Connection error' string to display the error, also explore errorMessages property
      };
      signals.newUserMessage.listener = (body) => {
        websocket.send(JSON.stringify(body));
      };
    } catch (e) {
      signals.onResponse({error: 'error'});
      signals.onClose();
      // reconnection logic here
    }
  },
}
```

**Error Handling**

All error handling must be done inside the `handler` and respond using the `onResponse({error: message})` signal.

**Release**

This is currently available in `deep-chat-dev` and `deep-chat-react-dev` version `9.0.5`. If all goes well, I am hoping to release this in the main `deep-chat` packages tomorrow.

Please send any feedback in regards to this.

**Comment by OvidijusParsiunas** [comment]
@GIOVANNI-75 in regards to the `requestInterceptor` and `onNewMessage` not working in the previous version, I could not  replicate the problem and it was working for me. By any chance, could you try the version `9.0.5` to see if they work for you there just in-case I am not catching some problem.

**Comment by OvidijusParsiunas** [comment]
Hey folks, the request [`handler`](https://deepchat.dev/docs/connect#Handler) functionality is now available in Deep Chat version `1.3.2` 🎉🎉🎉🎉

Thankyou very much for your guidance and critical feedback to get this feature out and without your efforts it would not exist today.

I have additionally released a [video](https://youtu.be/ySU7Tj99mbA?si=8kd6Q63AFFs_HQi5) to help introduce developers to this functionality.

Thankyou once again! If you experience any issues with this feature or have suggestions to improve it, feel free to add a comment here. For everything else, do not hesitate to create new issues. Thankyou!!!!!

**Comment by ryanisn** [comment]
It is a bit difficult to write the handler js functoin inside the connect json object in plain html w/ vanilla js.
Is it possible to add a onSubmmit event or beforeHandler listener to point to a js function?
